### PR TITLE
workdir and create_unique_workdir for ssh plugin

### PR DESCRIPTION
**Pull request by madhur-tandon**
Closes https://github.com/AgnostiqHQ/covalent-ssh-plugin/issues/64

- [ ] I have added the tests to cover my changes.
- [x] I have updated the documentation, VERSION, and CHANGELOG accordingly.
- [x] I have read the CONTRIBUTING document.

~~This is a WIP since we probably need to do the following:~~

~~- Have another variable corresponding to `remote_workdir` but for local machine i.e. not on the remote server~~
~~- Sync the `remote_workdir` contents into `workdir` (for local) if we do the above...~~

~~But I could appreciate some initial thoughts...~~

CC: @arosen93

### CODE REVIEW

**Comment by arosen93** [comment]
I'll take a look shortly but one initial comment that it might be worth renaming it to `remote_workdir` to match the convention that exists in the Slurm executor.

**Comment by arosen93** [comment]
What's the thought process behind needing a workdir for the local machine? The local machine shouldn't do anything here, right? The pickled results will still be stored in the `results_dir`

**Comment by madhur-tandon** [comment]
> What's the thought process behind needing a workdir for the local machine? The local machine shouldn't do anything here, right? The pickled results will still be stored in the `results_dir`.

What if you want the files produced in the remote machine's workdir synced to your local machine? Possibly preserving the same directory structure?

Again, this is something open ended and thus I mentioned about it in the first comment...

**Comment by arosen93** [comment]
That's an interesting option to consider. I think in that case it's probably best left to the user to set up the file transfer via one of Covalent's [file transfer options](https://covalent.readthedocs.io/en/latest/api/data_transfers.html). Just my personal take and is what @wjcunningham7 and I assumed in the Slurm executor.

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [UNRELEASED]
 
+### Added
+
+- Ability to specify a `workdir` along with `create_unique_workdir` option for each electron / node.
```
Comment (arosen93) [inline CHANGELOG.md:12]: Swap `workdir` to `remote_workdir`

File: covalent_ssh_plugin/ssh.py [hunk 2]
```diff
@@ -45,10 +45,12 @@
     "hostname": "",
     "ssh_key_file": os.path.join(os.environ["HOME"], ".ssh/id_rsa"),
     "cache_dir": str(Path(get_config("dispatcher.cache_dir")).expanduser().resolve()),
-    "remote_cache": ".cache/covalent",
     "python_path": "python",
-    "run_local_on_ssh_fail": False,
     "conda_env": "",
+    "remote_cache": ".cache/covalent",
+    "run_local_on_ssh_fail": False,
+    "remote_workdir": ".cache/covalent/remote_workdir",
```
Comment (arosen93) [inline covalent_ssh_plugin/ssh.py:52]: Probably better to change the `remote_workdir` default to `covalent-workdir` to match with the Slurm executor.

File: covalent_ssh_plugin/ssh.py [hunk 3]
```diff
@@ -99,30 +106,43 @@ def __init__(
         self.cache_dir = str(Path(self.cache_dir).expanduser().resolve())
 
         self.run_local_on_ssh_fail = run_local_on_ssh_fail
+
+        self.remote_workdir = remote_workdir or get_config("executors.ssh.remote_workdir")
+        self.create_unique_workdir = create_unique_workdir or get_config("executors.ssh.create_unique_workdir")
```
Comment (arosen93) [inline covalent_ssh_plugin/ssh.py:111]: If the user specified `create_unique_workdir = False`, it will be overriden and pulled from the config instead.

I think the more direct approach of           `get_config("executors.ssh.create_unique_workdir") if create_unique_workdir is None else create_unique_workdir` might be more robust, assuming the default is `None`.

**Review by arosen93 — COMMENTED** [review]
Looks great! Just left a few small comments to match consistency with the Slurm executor.

File: CHANGELOG.md (hunk 1 above)
Comment (madhur-tandon) [reply CHANGELOG.md]: missed it, thanks!

File: covalent_ssh_plugin/ssh.py (hunk 2 above)
Comment (madhur-tandon) [reply covalent_ssh_plugin/ssh.py]: done!

File: covalent_ssh_plugin/ssh.py (hunk 3 above)
Comment (madhur-tandon) [reply covalent_ssh_plugin/ssh.py]: done, the default value is `None` in the constructor but `False` in `_EXECUTOR_PLUGIN_DEFAULTS`

Comment (arosen93) [reply covalent_ssh_plugin/ssh.py]: Sweet, thanks for helping with the consistency overall! All makes sense to me!

**Comment by arosen93** [comment]
@madhur-tandon. Might be worth removing the WIP tag and doing anything related to file sync (much more detailed and nuanced discussion) to a future PR or issue. I think what you've done here looks good and should be merged as is once the team reviews and agrees!

**Comment by madhur-tandon** [comment]
@kessler-frost maybe you can review this as well perhaps...?

File: covalent_ssh_plugin/ssh.py [hunk 4]
```diff
@@ -79,6 +84,8 @@ def __init__(
         conda_env: str = None,
         remote_cache: str = "",
         run_local_on_ssh_fail: bool = False,
+        remote_workdir: str = "",
+        create_unique_workdir: Optional[bool] = None,
```
Comment (kessler-frost) [inline covalent_ssh_plugin/ssh.py:88]: If it's bool then let's make it's default to be `True` or `False`.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -43,7 +43,6 @@ repos:
     rev: 22.8.0
     hooks:
       - id: black
-        language_version: python3.8
```
Comment (kessler-frost) [inline .pre-commit-config.yaml:46]: Hmm, why was this removed?

**Review by kessler-frost — CHANGES_REQUESTED** [review]
Hey @madhur-tandon ! Thanks for this one as well. Just a some minor comments, otherwise this looks gtg 👍

File: .pre-commit-config.yaml (hunk 5 above)
Comment (madhur-tandon) [reply .pre-commit-config.yaml:46]: I removed this from other PRs as well, even in the main one..
The reason is, it expects `3.8` as the python version, else `black` throws an error...
I personally use `3.9.x` in my virtual environment..
So, it's just to use `pre-commit run --all` without necessarily having python 3.8.x

File: covalent_ssh_plugin/ssh.py (hunk 4 above)
Comment (madhur-tandon) [reply covalent_ssh_plugin/ssh.py:88]: The idea is to get it from the config i.e. `get_config("executors.ssh.create_unique_workdir")` if it's `None` i.e. it is not supplied. When the default is `False`, we don't know if it's `False` intentionally OR perhaps the default is `False` and it's an indication to get from the config...

`None` helps us in this case...

PS -- This is the same for `Local` and `Dask` executors as well...

Comment (kessler-frost) [reply covalent_ssh_plugin/ssh.py:88]: You're right, I was going to ask whether there's a specific reason. This seems reasonable 👍 .

File: covalent_ssh_plugin/ssh.py [hunk 6]
```diff
@@ -99,6 +106,14 @@ def __init__(
         self.cache_dir = str(Path(self.cache_dir).expanduser().resolve())
 
         self.run_local_on_ssh_fail = run_local_on_ssh_fail
+
+        self.remote_workdir = remote_workdir or get_config("executors.ssh.remote_workdir")
+        self.create_unique_workdir = (
+            get_config("executors.ssh.create_unique_workdir")
+            if create_unique_workdir is None
+            else create_unique_workdir
```
Comment (kessler-frost) [inline covalent_ssh_plugin/ssh.py:114]: (guess I somehow missed this part)

**Comment by kessler-frost** [comment]
@madhur-tandon seems like test coverage isn't up to the threshold. Feel free to ping me again once you've updated the tests 👍 .

File: covalent_ssh_plugin/ssh.py (hunk 6 above)
Comment (madhur-tandon) [reply covalent_ssh_plugin/ssh.py:114]: np, resolving this convo..

**Comment by madhur-tandon** [comment]
@kessler-frost requesting a re-review...

**Comment by kessler-frost** [comment]
@madhur-tandon Can you change the version of `isort` to `rev: 5.12.0` in `.pre-commit-config.yaml`. That should fix the pre-commit failing issue - that is the version we're using in main covalent as well.

**Review by kessler-frost — APPROVED** [review]
